### PR TITLE
[chore]: Refactor code

### PR DESCRIPTION
**Pull request by jeremyhager**
I noticed some of the code could be refactored, so I've organized some of the code into different files/modules, as well as added a few other things (listed below).
If there is a better idea for this or something has gone wrong please let me know. I hope to contribute some more along the way.

- `CommentData`, `Listings`, and `Submissions` derive `Default`
  - Easier to init structs with a few values
- `SubmissionData` and `Submission` have `::new()` functions for defaults
- Web configuration functions and structs in dedicated module/file
  - Errors, functions, etc. commonly used in the `web_conf` file
- `api_translation` takes advantage of defaults in some functions
- Endpoint info in dedicated module/file
  - Add default `"/"` route, at least for now
  - Default `"/"` route points back to this repo
- Move some pertinent readme info  from `tafkars/src/README.md` to `README.md`
- Tested locally for endpoint `/hot.json`

### CODE REVIEW

**Comment by derivator** [comment]
Thanks! Looks pretty good, going to be busy tomorrow, might have time for a proper review on sunday.  
I think my thinking when not deriving `Default` for those structs was that a default post doesn't really make sense. The [docs state](https://doc.rust-lang.org/std/default/trait.Default.html) that `Default` is "A trait for giving a type a useful default value." but an empty post is not really *useful*. However, from a practicality standpoint it totally makes sense, so maybe my thinking was just unnecessarily pedantic. It *probably* also makes sense to have `Default` derived from a `Deserialize` standpoint, though we don't use that here...

**Comment by jeremyhager** [comment]
When it comes to Defaults my thoughts were that it's easier to initialize structs this way, but also that we have a lot of unpopulated fields at the moment that are basically defaulting anyway. However, I don't feel that strongly about it so if you want to keep it the same that's good with me. The point that an empty post isn't useful is a good point in either case.

It may be better to instead use something like with Listings, where it takes a type - but we essentially have that when we pass to `posts` functions.

I'll try working on a Dockerfile if I have time and some GitHub actions, thanks for taking a look!

**Comment by derivator** [comment]
Been very busy, sorry for the late response. My current thoughts:

Agree on:
* `Default` impls
* endpoints.rs - probably also [implement configuring the routes here](https://docs.rs/actix-web/latest/actix_web/web/struct.ServiceConfig.html) so someone could import this as a library and reuse the configuration
* `/` endpoint

Unsure on:
* `Submission(Data)::new()` - not that useful and moreso than with Default, I feel like if it exists, it should produce valid values as defined by Reddit
* `Listing::new` - oh shit, I did that one :sweat_smile: - atm doesn't produce valid values, at least one of `after`/`before` should be set, dist/modhash should always be set as far as I can tell, actually not sure why they're optional.
* web_config.rs - I feel like `AppState`, `prepare` and `response_state` belong in `endpoints.rs`? Or maybe the name "config" is confusing me and it should be something else.

I actually don't feel that strongly about any of this, so let me know what you think and I might just be convinced to merge as-is :sweat_smile: 

I do kind of prefer to leave the readme as it is and just link to `tafkars-lemmy` directly. I want to keep them kind of separate. `tafkars` is the API definitions, `tafkars-lemmy` the proxy.

**Comment by lecoqjacob** [comment]
I want to help flesh out more of the api, but want to wait for this to merge before poking around!

**Comment by jeremyhager** [comment]
Apologies, made some updates:

> Agree on:
> 
> Default impls
> endpoints.rs - probably also [implement configuring the routes here](https://docs.rs/actix-web/latest/actix_web/web/struct.ServiceConfig.html) so someone could import this as a library and reuse the configuration
> / endpoint

Coolio, will leave as-is.

Updates:
- endpoints.rs - [implement configuring the routes here](https://docs.rs/actix-web/latest/actix_web/web/struct.ServiceConfig.html)
  - Updated
- Submission(Data)::new() - not that useful and moreso than with Default
  - True - I've removed the function. If we really want something like this we can always implement later on.
- Listing::new
  - I see this used in `comments_for_post`. For now what I've done is created another TODO on top of `Listing::new()` to require one of after/before and modhash/dist. If we'd rather get rid of it altogether, I think I can sort it out.
  - As for why they're _all_ optional...I suppose the og dev of roux must have figured one of them must be set, probably easier to just set them to optional.
- web_config.rs
  - My naming on this module was a bit of a poor choice. It should be `server_config.rs` or similar; it's a general config for the api server. I've renamed it, but I'm open to other names/ideas as well.
  - I renamed the error to `ServerSideError`. We could also call it "general error" or something
  - Moved `AppState`, `prepare` and `response_state` to `endpoints.rs`
- `tafkars` is the API definitions, `tafkars-lemmy` the proxy.
Gotcha - I've undone those changes. Retested `/hot.json`, seems to still respond okay.

File: tafkars-lemmy/src/api_translation.rs
Comment (derivator) [inline tafkars-lemmy/src/api_translation.rs:226]: Some of these fields shouldn't be removed: `Some(false)` is different from `None`. RedReader fails to parse the response like this.

File: tafkars-lemmy/src/api_translation.rs
Comment (derivator) [inline tafkars-lemmy/src/api_translation.rs:259]: These as well.

**Review by derivator — CHANGES_REQUESTED** [review]
Will merge after this is fixed.

**Comment by jeremyhager** [comment]
> Will merge after this is fixed.
Got it, just pushed a few changes.

**Comment by jeremyhager** [comment]
wait a sec, just realized I need to import MaybeReplies

File: tafkars-lemmy/src/api_translation.rs
Comment (jeremyhager) [reply tafkars-lemmy/src/api_translation.rs]: Changed back

File: tafkars-lemmy/src/api_translation.rs
Comment (jeremyhager) [reply tafkars-lemmy/src/api_translation.rs]: changed back

**Comment by jeremyhager** [comment]
Nevermind, vscode is just being weird. Should be good to go

**Comment by derivator** [comment]
Sorry I wasn't being clear, the `None`s can go as they will correctly bet set from `Default`, just the `Some` values should stay.

**Comment by jeremyhager** [comment]
> Sorry I wasn't being clear, the `None`s can go as they will correctly bet set from `Default`, just the `Some` values should stay.

My bad, I think I see what you mean now.

**Comment by jeremyhager** [comment]
yay, thanks!

**Comment by derivator** [comment]
Thanks for your contribution @jeremyhager! Looking forward to more! :)

**Pull request merged**
